lib/linked_list.rb

```diff
@@ -7,39 +7,69 @@ def initialize
     @head = nil
   end
 
-  # Time complexity - ?
-  # Space complexity - ?
+  # Time complexity - o(1)?
+  # Space complexity -o(1) ?
   def add_first(data)
-
+    @head = Node.new(data, @head)
   end
 
-  # Time complexity - ?
-  # Space complexity - ?
+  # Time complexity - o(1) ?
+  # Space complexity - o(1)
   def get_first
-
+    if head
+      return head.data
+    else
+      return nil
+    end
   end
 
-  # Time complexity - ?
-  # Space complexity - ?
+  # Time complexity - o(n)?
+  # Space complexity - o(1) ?
   def length
-    return 0
+    current = head
+    length = 0
+    while current != nil
+      length += 1
+      current = current.next
+    end
+    return length
   end
 
-  # Time complexity - ?
-  # Space complexity - ?
+  # Time complexity - o(n) ?
+  # Space complexity - o(1)?
   def add_last(data)
-
+    current = head
+    if current == nil
+      self.add_first(data)
+      return
+    end
+    while current.next != nil
+      current = current.next
+    end
+    current.next = Node.new(data)
   end
 
-  # Time complexity - ?
-  # Space complexity - ?
+  # Time complexity - o(n)
+  # Space complexity - o(1)
   def get_last
-
+    current = head
```

Comment:
What if `head` is `nil`?

```suggestion
    return nil if head.nil?
    current = head
```